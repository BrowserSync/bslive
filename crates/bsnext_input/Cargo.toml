[package]
name = "bsnext_input"
version = "0.2.3"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bsnext_resp = { path = "../bsnext_resp" }
bsnext_tracing = { path = "../bsnext_tracing" }
bsnext_guards = { path = "../bsnext_guards" }

miette = { workspace = true }

random_word = { workspace = true }
matchit = { workspace = true }
mime_guess = { workspace = true }
serde = { workspace = true }
clap = { workspace = true }
anyhow = { workspace = true }
serde_yaml = { workspace = true }
serde_json = { workspace = true }
toml = { workspace = true }
thiserror = { workspace = true }

[dev-dependencies]
insta = { workspace = true }
