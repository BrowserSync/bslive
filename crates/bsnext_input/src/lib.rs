use crate::target::TargetKind;

use std::fmt::{Display, Formatter};
use std::fs;
use std::fs::read_to_string;
use std::net::AddrParseError;
use std::path::{Path, PathBuf};

use crate::md::MarkdownError;
use crate::yml::YamlError;

#[cfg(test)]
pub mod input_test;
pub mod md;
pub mod paths;
pub mod route;
pub mod route_manifest;
pub mod server_config;
pub mod target;
#[cfg(test)]
pub mod watch_opt_test;
pub mod yml;

#[derive(Debug, Default, serde::Deserialize, serde::Serialize)]
pub struct Input {
    pub servers: Vec<server_config::ServerConfig>,
}

impl Input {
    pub fn from_input_path<P: AsRef<Path>>(path: P) -> Result<Self, InputError> {
        match path.as_ref().extension().and_then(|x| x.to_str()) {
            None => Err(InputError::MissingExtension(path.as_ref().to_owned())),
            Some("yml") | Some("yaml") => Input::from_yaml_path(path),
            Some("md") | Some("markdown") => Input::from_md_path(path),
            Some(other) => Err(InputError::UnsupportedExtension(other.to_string())),
        }
    }
    fn from_yaml_path<P: AsRef<Path>>(path: P) -> Result<Self, InputError> {
        let str = read_to_string(&path)?;
        let output = serde_yaml::from_str::<Self>(str.as_str()).map_err(move |e| {
            if let Some(location) = e.location() {
                YamlError::ParseErrorWithLocation {
                    serde_error: e,
                    input: str,
                    path: path.as_ref().to_string_lossy().to_string(),
                    line: location.line(),
                    column: location.column(),
                }
            } else {
                YamlError::ParseError {
                    serde_error: e,
                    input: str,
                    path: path.as_ref().to_string_lossy().to_string(),
                }
            }
        })?;
        // todo: don't allow duplicates?.
        Ok(output)
    }
    fn from_md_path<P: AsRef<Path>>(path: P) -> Result<Self, InputError> {
        let str = read_to_string(path)?;
        let input = md::md_to_input(&str)?;
        // todo: don't allow duplicates.
        Ok(input)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum InputError {
    #[error("no suitable inputs could be found")]
    MissingInputs,
    #[error("could not read input, error: {0}")]
    InvalidInput(String),
    #[error("io error")]
    Io(#[from] std::io::Error),
    #[error("Could not find the input file: {0}")]
    NotFound(PathBuf),
    #[error("Paths without extensions are not supported: {0}")]
    MissingExtension(PathBuf),
    #[error("Unsupported extension: {0}")]
    UnsupportedExtension(String),
    #[error("InputWriteError prevented startup {0}")]
    InputWriteError(#[from] InputWriteError),
    #[error("Input path error prevented startup {0}")]
    PathError(#[from] PathError),
    #[error("Input port error prevented startup {0}")]
    PortError(#[from] PortError),
    #[error("Input directory error prevented startup {0}")]
    DirError(#[from] DirError),
    #[error("Markdown error: {0}")]
    MarkdownError(#[from] MarkdownError),
    #[error("{0}")]
    YamlError(#[from] YamlError),
}

#[derive(Debug, thiserror::Error)]
pub enum WatchError {
    #[error("don't add `.` before the extension.")]
    InvalidExtensionFilter,
    #[error("empty")]
    EmptyExtensionFilter,
}

#[derive(Debug, thiserror::Error)]
pub enum InputWriteError {
    #[error("couldn't write input to {path}")]
    FailedWrite { path: PathBuf },
}

#[derive(Debug, thiserror::Error, serde::Serialize, serde::Deserialize)]
pub enum PathError {
    #[error("path(s) not found \n{paths}")]
    MissingPaths { paths: PathDefs },
}

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub enum PortError {
    #[error("could not use that port: {port} {err}")]
    InvalidPort { port: u16, err: AddrParseError },
}

#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub enum DirError {
    #[error("could not create that dir: {path}")]
    CannotCreate { path: PathBuf },
    #[error("could not change the process CWD to: {path}")]
    CannotMove { path: PathBuf },
}

#[derive(Debug, thiserror::Error, serde::Serialize, serde::Deserialize)]
pub struct PathDefs(Vec<PathDefinition>);

#[derive(Debug, thiserror::Error, serde::Serialize, serde::Deserialize)]
struct PathDefinition {
    pub input: String,
    pub cwd: PathBuf,
    pub absolute: PathBuf,
}

impl Display for PathDefs {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        for pd in self.0.iter() {
            writeln!(f, "  cwd:   {}", pd.cwd.display())?;
            writeln!(f, "  input: {}", pd.input)?;
        }
        Ok(())
    }
}
impl Display for PathDefinition {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("PathDefinition").finish()
    }
}

pub fn rand_word() -> String {
    random_word::gen(random_word::Lang::En).to_string()
}

pub fn fs_write_input(
    cwd: &Path,
    input: &Input,
    target_kind: TargetKind,
) -> Result<PathBuf, InputWriteError> {
    let string = match target_kind {
        TargetKind::Yaml => serde_yaml::to_string(&input).expect("create yaml?"),
        TargetKind::Toml => toml::to_string_pretty(&input).expect("create toml?"),
        TargetKind::Md => md::input_to_str(input),
    };
    let name = match target_kind {
        TargetKind::Yaml => "input.yml",
        TargetKind::Toml => "input.toml",
        TargetKind::Md => "input.md",
    };
    let next_path = cwd.join(name);
    fs::write(&next_path, string)
        .map(|()| next_path.clone())
        .map_err(|_e| InputWriteError::FailedWrite { path: next_path })
}
