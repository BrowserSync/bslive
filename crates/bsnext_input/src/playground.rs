use crate::path_def::PathDef;
use crate::route::{FallbackRoute, Opts, Route, RouteKind};
use bsnext_resp::builtin_strings::{BuiltinStringDef, BuiltinStrings};
use bsnext_resp::inject_addition::{AdditionPosition, InjectAddition};
use bsnext_resp::inject_opts::{InjectOpts, Injection, InjectionItem, MatcherList};
use bsnext_resp::path_matcher::PathMatcher;

#[derive(Debug, PartialEq, Default, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct Playground {
    pub html: String,
    pub js: Option<String>,
    pub css: Option<String>,
}

const FALLBACK_HTML: &str = "This is a BSLIVE playground";

impl Playground {
    pub fn as_routes(&self) -> Vec<Route> {
        let home_path = PathDef::try_new("/");
        let js_path = PathDef::try_new("/__bslive_playground.js");
        let css_path = PathDef::try_new("/__bslive_playground.css");
        let reset_path = PathDef::try_new("/reset.css");

        let (Ok(home), Ok(js), Ok(css), Ok(reset_path)) =
            (home_path, js_path, css_path, reset_path)
        else {
            return vec![];
        };

        let home_route = Route {
            path: home,
            kind: RouteKind::new_html(self.html.clone()),
            opts: Opts {
                cors: None,
                delay: None,
                watch: Default::default(),
                inject: playground_wrap(),
                headers: None,
                compression: None,
                ..Default::default()
            },
            fallback: Some(FallbackRoute {
                kind: RouteKind::new_html(FALLBACK_HTML),
                opts: Default::default(),
            }),
        };
        let js_route = Route {
            path: js,
            kind: RouteKind::new_raw(
                self.js
                    .as_ref()
                    .unwrap_or(&"// playground js is absent".to_string()),
            ),
            ..Default::default()
        };
        let css_route = Route {
            path: css,
            kind: RouteKind::new_raw(
                self.css
                    .as_ref()
                    .unwrap_or(&"/* playground css is absent */".to_string()),
            ),
            ..Default::default()
        };
        let reset_route = Route {
            path: reset_path,
            kind: RouteKind::new_raw(include_str!("../../../ui/styles/reset.css")),
            ..Default::default()
        };
        vec![home_route, js_route, css_route, reset_route]
    }
}

fn playground_wrap() -> InjectOpts {
    let prepend = r#"
<!doctype html>
  <html lang="en">
  <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Document</title>
      <link rel="stylesheet" href="/__bslive_playground.css">
  </head>
  <body>
"#;
    let append = r#"
  <script src="/__bslive_playground.js"></script>
  </body>
</html>
"#;
    InjectOpts::Items(vec![
        InjectionItem {
            inner: Injection::Addition(InjectAddition {
                addition_position: AdditionPosition::Prepend(prepend.to_string()),
            }),
            only: Some(MatcherList::Item(PathMatcher::Str("/".to_string()))),
        },
        InjectionItem {
            inner: Injection::Addition(InjectAddition {
                addition_position: AdditionPosition::Append(append.to_string()),
            }),
            only: Some(MatcherList::Item(PathMatcher::Str("/".to_string()))),
        },
        InjectionItem {
            inner: Injection::BsLive(BuiltinStringDef {
                name: BuiltinStrings::Connector,
            }),
            only: Some(MatcherList::Item(PathMatcher::Str("/".to_string()))),
        },
    ])
}
