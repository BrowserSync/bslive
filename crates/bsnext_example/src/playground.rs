use bsnext_input::server_config::ServerIdentity;
use bsnext_input::{InputCreation, InputCtx, InputSource, InputSourceKind};
use bsnext_md::md_fs::MdFs;

#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct PlaygroundExample;

impl InputSource for PlaygroundExample {
    fn into_input(self, identity: Option<ServerIdentity>) -> InputSourceKind {
        let input_str = include_str!("../../../examples/markdown/playground.md");
        let mut input = MdFs::from_input_str(input_str, &InputCtx::default()).unwrap();

        // update the server identity if it was provided
        if let (Some(server), Some(identity)) = (input.servers.get_mut(0), identity) {
            server.identity = identity;
        }

        InputSourceKind::Type(input)
    }
}
