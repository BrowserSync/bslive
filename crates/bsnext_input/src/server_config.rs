use crate::client_config::ClientConfig;
use crate::playground::Playground;
use crate::route::{Route, Watcher};
use crate::{rand_word, PortError};
use serde::{de, Deserializer};
use std::hash::{DefaultHasher, Hash, Hasher};
use std::net::SocketAddr;
use std::str::FromStr;

#[derive(Debug, Default, Clone, serde::Deserialize, serde::Serialize)]
pub struct ServerConfig {
    #[serde(flatten)]
    pub identity: ServerIdentity,
    #[serde(default)]
    pub routes: Vec<Route>,
    #[serde(default)]
    pub watchers: Vec<Watcher>,
    #[serde(default)]
    pub playground: Option<Playground>,
    #[serde(default)]
    pub clients: ClientConfig,
}

impl ServerConfig {
    ///
    /// All regular routes, plus dynamically added ones (for example, through a playground)
    ///
    pub fn combined_routes(&self) -> Vec<Route> {
        let routes = self.routes.clone();
        match &self.playground {
            None => self.routes.clone(),
            Some(playground) => {
                let mut pg_routes = playground.as_routes();
                pg_routes.extend(routes);
                pg_routes
            }
        }
    }
    pub fn raw_routes(&self) -> &[Route] {
        &self.routes
    }

    pub fn from_route(r: Route) -> Self {
        Self {
            routes: vec![r],
            ..std::default::Default::default()
        }
    }
}

#[derive(
    Debug, Ord, PartialOrd, PartialEq, Hash, Eq, Clone, serde::Deserialize, serde::Serialize,
)]
#[serde(untagged)]
pub enum ServerIdentity {
    PortNamed {
        #[serde(deserialize_with = "deserialize_port")]
        port: u16,
        name: String,
    },
    Both {
        name: String,
        bind_address: String,
    },
    Address {
        bind_address: String,
    },
    Named {
        name: String,
    },
    Port {
        #[serde(deserialize_with = "deserialize_port")]
        port: u16,
    },
}

fn deserialize_port<'de, D>(deserializer: D) -> Result<u16, D::Error>
where
    D: Deserializer<'de>,
{
    struct PortVisitor;

    impl<'de> de::Visitor<'de> for PortVisitor {
        type Value = u16;

        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
            formatter.write_str("a string or integer representing a valid port number")
        }

        fn visit_u64<E>(self, value: u64) -> Result<u16, E>
        where
            E: de::Error,
        {
            if value <= u16::MAX as u64 {
                Ok(value as u16)
            } else {
                Err(E::custom(format!("port number out of range: {}", value)))
            }
        }

        fn visit_str<E>(self, value: &str) -> Result<u16, E>
        where
            E: de::Error,
        {
            value
                .parse::<u16>()
                .map_err(|_| E::custom(format!("invalid port number: {}", value)))
        }
    }

    deserializer.deserialize_any(PortVisitor)
}

impl Default for ServerIdentity {
    fn default() -> Self {
        Self::named()
    }
}

impl ServerIdentity {
    pub fn named() -> Self {
        Self::Named { name: rand_word() }
    }
    pub fn address<A: AsRef<str>>(a: A) -> Self {
        Self::Address {
            bind_address: a.as_ref().to_string(),
        }
    }
    /// if a port was provided, try to use it by validating it first
    /// otherwise default to a named identity
    pub fn from_port_or_named(port: Option<u16>) -> Result<Self, PortError> {
        let result = port.map(|port| {
            SocketAddr::from_str(&format!("0.0.0.0:{port}"))
                .map_err(|err| PortError::InvalidPort { port, err })
        });
        match result {
            None => Ok(ServerIdentity::named()),
            Some(Ok(addr)) => Ok(ServerIdentity::address(addr.to_string())),
            Some(Err(err)) => Err(err),
        }
    }

    pub fn as_id(&self) -> u64 {
        let mut hasher = DefaultHasher::new();
        self.hash(&mut hasher);
        hasher.finish()
    }

    pub fn is_named(&self, pred_name: &str) -> bool {
        match self {
            ServerIdentity::Both { name, .. } => name == pred_name,
            ServerIdentity::Address { .. } => false,
            ServerIdentity::Named { name } => name == pred_name,
            ServerIdentity::Port { .. } => false,
            ServerIdentity::PortNamed { name, .. } => name == pred_name,
        }
    }
}

#[test]
fn server_config_as_enum() {
    #[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
    struct C {
        servers: Vec<ServerConfig>,
    }
    let input = r#"
servers:
    - bind_address: 127.0.0.1:3000
    - name: server_1
    - name: server_2
      bind_address: 127.0.0.1:3001
      routes:
        - path: /
          dir: .
    "#;
    let c: C = serde_yaml::from_str(input).unwrap();

    // let baseline = ServerConfigInput::Named { name: "server_1".into() };
    let baseline = ServerIdentity::Address {
        bind_address: "127.0.0.1x:3000".into(),
    };
    for x in &c.servers {
        assert_eq!(x.identity == baseline, false);
    }
}

#[test]
fn with_port() {
    #[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
    struct C {
        servers: Vec<ServerConfig>,
    }
    let input = r#"
servers:
    - port: 3000
    - port: "3001"
    "#;
    let c: C = serde_yaml::from_str(input).unwrap();
    let expected = ServerIdentity::Port { port: 3000 };
    let first = c.servers.first().unwrap();
    assert_eq!(first.identity, expected);

    let second = c.servers.get(1).unwrap();
    let second_expected = ServerIdentity::Port { port: 3001 };
    assert_eq!(second.identity, second_expected);
}
