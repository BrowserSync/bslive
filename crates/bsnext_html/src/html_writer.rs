use bsnext_input::{Input, InputWriter};

pub struct HtmlWriter;

impl InputWriter for HtmlWriter {
    fn input_to_str(&self, input: &Input) -> String {
        if input.servers.is_empty() {
            todo!("html requires at least 1 server definition")
        }
        if input.servers.len() > 1 {
            todo!("more than 1 server not supported yet")
        }
        let server = input.servers.first().expect("must access first");
        let Some(playground) = &server.playground else {
            todo!("only playground is supported in HTML for now")
        };
        let mut blocks: Vec<String> = vec![];
        if let Some(css) = &playground.css {
            blocks.push("<style>".into());
            let indented = indent::indent_all_by(4, css);
            blocks.push(indented);
            blocks.push("</style>".into());
        }
        blocks.push(playground.html.clone());
        if let Some(js) = &playground.js {
            blocks.push("<script type=\"module\">".into());
            let indented = indent::indent_all_by(4, js);
            blocks.push(indented);
            blocks.push("</script>".into());
        }
        blocks.join("\n")
    }
}

#[test]
fn test_html_writer_for_playground() {
    use bsnext_input::playground::Playground;
    use bsnext_input::server_config::ServerConfig;
    let css = r#"body {
    background: red;
}"#;
    let js = r#"console.log("hello world!")"#;
    let playground = Playground {
        html: "<p>Hello world</p>".to_string(),
        js: Some(js.to_string()),
        css: Some(css.to_string()),
    };
    let mut input = Input::default();
    let mut server = ServerConfig::default();
    server.playground = Some(playground);
    input.servers.push(server);
    let output = HtmlWriter.input_to_str(&input);
    insta::assert_snapshot!(output);
}
