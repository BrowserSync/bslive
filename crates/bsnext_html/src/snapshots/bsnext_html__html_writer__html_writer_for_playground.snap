---
source: crates/bsnext_html/src/html_writer.rs
expression: output
---
<style>
    body {
        background: red;
    }
</style>
<p>Hello world</p>
<script type="module">
    console.log("hello world!")
</script>
